File: crates/mogeungd/src/notify.rs
//! Telling you a session needs attention when you are not looking at mogeung.
//!
//! Roadmap `R-C1` (macOS notification) and `R-C4` (push to a phone).
//!
//! The hard part is not delivery, it is **not being annoying**. A notifier that
//! fires on every scan trains you to dismiss it, and then the one that mattered
//! gets dismissed too — the same failure the format canary had to learn (see
//! `docs/design/health-and-canary.md`).
//!
//! So the rule here is: notify on the *transition* into needing you, once, per
//! session. Never on a state that is merely continuing.

use mogeung_core::attention::{AttentionItem, AttentionReason};
use mogeung_core::session::SessionId;
use std::collections::HashMap;

/// Where a notification should go.
#[derive(Debug, Clone, Default)]
pub struct NotifyConfig {
    /// Post a desktop banner: osascript on macOS, notify-send on Linux.
    pub desktop: bool,
    /// POST the message to this URL (ntfy.sh style: the body *is* the message).
    pub push_url: Option<String>,
}

impl NotifyConfig {
    pub fn enabled(&self) -> bool {
        self.desktop || self.push_url.is_some()
    }
}

/// One thing worth telling the user, already rendered.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Notification {
    pub session_id: SessionId,
    pub title: String,
    pub body: String,
}

/// Remembers what it has already said, so it does not repeat itself.
#[derive(Default)]
pub struct Notifier {
    /// Last reason we notified about, per session.
    announced: HashMap<SessionId, AttentionReason>,
    cfg: NotifyConfig,
}

impl Notifier {
    pub fn new(cfg: NotifyConfig) -> Self {
        Notifier {
            announced: HashMap::new(),
            cfg,
        }
    }

    pub fn config(&self) -> &NotifyConfig {
        &self.cfg
    }

    /// Decide what to announce for the current queue.
    ///
    /// Pure: it returns the notifications and updates the seen-state, but sends
    /// nothing. That keeps the interesting logic — *when do we speak?* —
    /// testable without a desktop or a network.
    pub fn diff(&mut self, queue: &[AttentionItem], label_of: impl Fn(&str) -> String) -> Vec<Notification> {
        let mut out = Vec::new();
        let mut seen: HashMap<SessionId, AttentionReason> = HashMap::new();

        for item in queue {
            seen.insert(item.session_id.clone(), item.reason);

            if !item.reason.needs_human() {
                continue;
            }
            // Already told you about this exact state.
            if self.announced.get(&item.session_id) == Some(&item.reason) {
                continue;
            }

            let title = match item.reason {
                AttentionReason::AwaitingPermission => "Needs your approval",
                AttentionReason::AwaitingInput => "Waiting for you",
                AttentionReason::Failed => "Session failed",
                AttentionReason::NeedsReview => "Changes to review",
                AttentionReason::Stalled => "Session has gone quiet",
                _ => continue,
            };
            out.push(Notification {
                session_id: item.session_id.clone(),
                title: title.to_string(),
                body: format!("{} — {}", label_of(&item.session_id), item.detail),
            });
        }

        // Sessions that dropped out of the queue entirely are forgotten, so
        // that a session which needs you *again* later is announced again.
        self.announced = seen;
        out
    }

    /// Deliver. Best-effort and non-blocking-ish: a failed notification must
    /// never disturb the scan loop.
    pub fn send(&self, n: &Notification) {
        if self.cfg.desktop {
            // Runtime-selected (`cfg!`, not `#[cfg]`) so both platforms'
            // argument builders compile — and stay testable — everywhere.
            if cfg!(target_os = "macos") {
                let script = format!(
                    "display notification {} with title {}",
                    applescript_string(&n.body),
                    applescript_string(&format!("mogeung — {}", n.title))
                );
                let _ = std::process::Command::new("osascript")
                    .arg("-e")
                    .arg(script)
                    .status();
            } else {
                // notify-send takes arguments, not code: no shell is involved
                // and nothing is escaped, because nothing is interpreted.
                match std::process::Command::new("notify-send")
                    .args(notify_send_args(&n.title, &n.body))
                    .spawn()
                {
                    // Reaped off-thread so a slow notification daemon cannot
                    // stall the scan loop, and no zombie is left behind.
                    Ok(mut child) => {
                        std::thread::spawn(move || {
                            let _ = child.wait();
                        });
                    }
                    // notify-send absent (a headless box) is ordinary.
                    Err(_) => {}
                }
            }
        }
        if let Some(url) = &self.cfg.push_url {
            // curl rather than an HTTP client dependency: this is one POST on a
            // rare event, and shelling out cannot poison the async runtime.
            let _ = std::process::Command::new("curl")
                .args([
                    "-fsS",
                    "-m",
                    "10",
                    "-H",
                    &format!("Title: mogeung: {}", n.title),
                    "-d",
                    &n.body,
                    url,
                ])
                .output();
        }
    }
}

/// The argv for a Linux desktop banner. Pure, so the hostile-title test can
/// pin the shape without a desktop.
///
/// The `--` matters: titles come from session labels and transcript detail,
/// and a label starting with `-` would otherwise be read as an option. After
/// it, title and body ride as two verbatim arguments — notify-send has no
/// code to inject into, which is the whole safety argument here.
fn notify_send_args(title: &str, body: &str) -> Vec<String> {
    vec![
        "--app-name=mogeung".to_string(),
        "--".to_string(),
        format!("mogeung — {title}"),
        body.to_string(),
    ]
}

/// Quote a string for AppleScript. Backslashes first, then quotes.
fn applescript_string(s: &str) -> String {
    let escaped = s.replace('\\', "\\\\").replace('"', "\\\"");
    // Newlines inside an AppleScript literal are a syntax error.
    let flat = escaped.replace('\n', " ").replace('\r', " ");
    format!("\"{flat}\"")
}

#[cfg(test)]
mod tests {
    use super::*;

    fn item(id: &str, reason: AttentionReason) -> AttentionItem {
        AttentionItem {
            session_id: id.into(),
            reason,
            score: reason.base_score(),
            detail: "because".into(),
            since: None,
        }
    }

    fn label(id: &str) -> String {
        format!("session {id}")
    }

    #[test]
    fn announces_a_transition_into_needing_you() {
        let mut n = Notifier::default();
        let out = n.diff(&[item("a", AttentionReason::AwaitingInput)], label);
        assert_eq!(out.len(), 1);
        assert_eq!(out[0].title, "Waiting for you");
        assert!(out[0].body.contains("session a"));
    }

    /// The rule that makes this usable. Without it, every 1.5s scan re-announces
    /// every waiting session and you turn notifications off within a minute.
    #[test]
    fn does_not_repeat_itself_while_the_state_persists() {
        let mut n = Notifier::default();
        let q = vec![item("a", AttentionReason::AwaitingInput)];
        assert_eq!(n.diff(&q, label).len(), 1);
        for _ in 0..20 {
            assert!(n.diff(&q, label).is_empty());
        }
    }

    #[test]
    fn a_change_of_reason_is_worth_saying() {
        let mut n = Notifier::default();
        assert_eq!(n.diff(&[item("a", AttentionReason::AwaitingInput)], label).len(), 1);
        // Escalated from "waiting for a task" to "blocked on approval".
        let out = n.diff(&[item("a", AttentionReason::AwaitingPermission)], label);
        assert_eq!(out.len(), 1);
        assert_eq!(out[0].title, "Needs your approval");
    }

    #[test]
    fn going_quiet_then_needing_you_again_is_announced_again() {
        let mut n = Notifier::default();
        assert_eq!(n.diff(&[item("a", AttentionReason::AwaitingInput)], label).len(), 1);
        // Back to work: nothing to say.
        assert!(n.diff(&[item("a", AttentionReason::Running)], label).is_empty());
        // Waiting once more — this is new information.
        assert_eq!(n.diff(&[item("a", AttentionReason::AwaitingInput)], label).len(), 1);
    }

    #[test]
    fn sessions_that_do_not_need_you_are_never_announced() {
        let mut n = Notifier::default();
        let q = vec![
            item("a", AttentionReason::Running),
            item("b", AttentionReason::Idle),
        ];
        assert!(n.diff(&q, label).is_empty());
    }

    /// notify-send receives arguments, not code — so a hostile title must
    /// arrive *verbatim*, one string, inside a single argv slot. Any escaping,
    /// splitting, or shell in the path would show up here as a mismatch.
    #[test]
    fn notify_send_gets_hostile_text_verbatim_with_no_shell() {
        let title = r#"$(rm -rf /) `touch /pwned` "quoted" 'single'"#;
        let body = r#"`id` $(reboot) ; rm -rf / # "all" of it"#;
        let args = notify_send_args(title, body);

        assert_eq!(args.len(), 4, "app-name, --, title, body — nothing more");
        assert_eq!(args[0], "--app-name=mogeung");
        assert_eq!(args[1], "--", "a title starting with `-` must not become an option");
        assert_eq!(args[2], format!("mogeung — {title}"), "title verbatim, one arg");
        assert_eq!(args[3], body, "body verbatim, one arg");

        // No argument is a shell or asks one to interpret anything.
        for a in &args {
            assert_ne!(a, "sh");
            assert_ne!(a, "bash");
            assert_ne!(a, "-c");
        }
        // The hostile text was not escaped into something else: exactly one
        // argument contains each payload, character for character.
        assert_eq!(args.iter().filter(|a| a.contains("$(rm -rf /)")).count(), 1);
        assert_eq!(args.iter().filter(|a| a.contains("`touch /pwned`")).count(), 1);
    }

    #[test]
    fn applescript_quoting_survives_a_hostile_prompt() {
        let s = applescript_string("he said \"hi\\bye\"\nand left");
        assert!(!s[1..s.len() - 1].contains('\n'));
        assert_eq!(s, r#""he said \"hi\\bye\" and left""#);
    }
}

File: desktop/src/lib/notify.ts
/**
 * Telling you a session needs you when you are not looking. `R-C1`.
 *
 * A port of `crates/mogeungd/src/notify.rs`, and deliberately a faithful one:
 * delivery is the easy half, and the hard half — **not being annoying** — was
 * reasoned out once there and should not be re-derived differently here. The
 * rule is the transition *into* needing you, once, per session; never a state
 * that is merely continuing, or every snapshot would re-announce every waiting
 * session.
 *
 * ## Who speaks
 *
 * Two processes can post the same banner, and one of them must not. `mogeungd`
 * notifies when it is run with `--notify`; a **hosted** daemon is started with
 * it off, on the reasoning written into `src-tauri/src/daemon.rs` — the window
 * is right there. That leaves the case nobody covers: this window hosting a
 * daemon while you are working in another application, which is precisely when
 * a queue nobody can see is worth interrupting for.
 *
 * So the window speaks only when it hosts, and only while it is unfocused. A
 * daemon you attached to is someone else's to announce, and if you want banners
 * from a daemon that outlives the window, `mogeungd --notify` is the thing that
 * gives them to you.
 */

import { useEffect, useRef } from "react";
import type { AttentionItem, AttentionReason, SessionId } from "@/wire/types";
import { needsHuman, sessionLabel } from "@/wire/types";
import { isTauri } from "@/lib/tauri";
import { useStore } from "@/store";

export interface Announcement {
  sessionId: SessionId;
  title: string;
  body: string;
}

/** The same words the daemon uses, so the two cannot describe one state twice. */
function titleFor(reason: AttentionReason): string | null {
  switch (reason) {
    case "awaiting_permission":
      return "Needs your approval";
    case "awaiting_input":
      return "Waiting for you";
    case "failed":
      return "Session failed";
    case "needs_review":
      return "Changes to review";
    case "stalled":
      return "Session has gone quiet";
    default:
      // `rate_limited` counts as needing you but has no line here, exactly as
      // in the Rust: the queue shows it, and there is nothing you can do about
      // a limit by being interrupted.
      return null;
  }
}

/**
 * What to say about this queue, given what has already been said.
 *
 * Pure, like `Notifier::diff`: it returns the announcements *and* the next
 * seen-state and sends nothing, which is what makes the interesting question —
 * when do we speak? — testable with no desktop and no permissions.
 *
 * `next` is rebuilt from the queue rather than merged into, so a session that
 * drops out entirely is forgotten and is announced again if it comes back
 * needing you later.
 */
export function diffQueue(
  announced: ReadonlyMap<SessionId, AttentionReason>,
  queue: readonly AttentionItem[],
  labelOf: (id: SessionId) => string,
): { out: Announcement[]; next: Map<SessionId, AttentionReason> } {
  const out: Announcement[] = [];
  const next = new Map<SessionId, AttentionReason>();

  for (const item of queue) {
    next.set(item.session_id, item.reason);
    if (!needsHuman(item.reason)) continue;
    // Already told you about this exact state.
    if (announced.get(item.session_id) === item.reason) continue;
    const title = titleFor(item.reason);
    if (!title) continue;
    out.push({
      sessionId: item.session_id,
      title,
      body: `${labelOf(item.session_id)} — ${item.detail}`,
    });
  }

  return { out, next };
}

/**
 * Ask the OS, once, and remember the answer for this run.
 *
 * Requested when you turn banners on rather than at startup: a permission
 * prompt on first launch, for a feature that is off by default, is exactly the
 * overstep the daemon's `--notify` flag exists to avoid.
 */
export async function ensurePermission(): Promise<boolean> {
  if (!isTauri()) return false;
  const { isPermissionGranted, requestPermission } = await import(
    "@tauri-apps/plugin-notification"
  );
  if (await isPermissionGranted()) return true;
  return (await requestPermission()) === "granted";
}

/** Post one banner. Best-effort: a notifier that throws must not break a scan. */
export async function deliver(n: Announcement): Promise<void> {
  if (!isTauri()) return;
  try {
    const { sendNotification } = await import("@tauri-apps/plugin-notification");
    sendNotification({ title: `mogeung — ${n.title}`, body: n.body });
  } catch {
    // A refused or unavailable notification service is ordinary — the same
    // posture `notify.rs` takes with a missing `notify-send`.
  }
}

/**
 * What the bell is actually going to do, in three states rather than two.
 *
 * Asked on 2026-08-07 — *"I tried to enable it, but I didn't see any behaviour
 * changes"* — and the honest answer was that in the usual setup the toggle
 * **cannot** do anything: `start.sh` runs the daemon as its own process, so the
 * window is *attached*, and an attached window deliberately stays silent
 * (see the note at the top of this file). The tooltip said so; the icon did
 * not, and a control that looks switched on while being unable to act is a
 * control that teaches you not to trust it.
 *
 * - `off` — nothing will be posted.
 * - `here` — this window hosts the daemon and will post, while unfocused.
 * - `elsewhere` — banners are on, but the **daemon** is the one that announces.
 *   Which it does when run with `--notify`, as `scripts/start.sh` does by
 *   default. Nothing is broken; this window simply is not the speaker.
 */
export type BellState = "off" | "here" | "elsewhere";

export function bellState(notify: boolean, hosting: boolean): BellState {
  if (!notify) return "off";
  return hosting ? "here" : "elsewhere";
}

/**
 * Watch the queue and announce what changes. Called once, by `App`.
 *
 * The seen-state is advanced **even while silent**, which is the difference
 * between a notifier and a backlog: without it, turning banners on — or coming
 * back to the window after an hour — would fire one for every session that has
 * been waiting all along, and the first thing you would do is turn them off
 * again.
 */
export function useNotifications(): void {
  const queue = useStore((s) => s.queue);
  const sessions = useStore((s) => s.sessions);
  const enabled = useStore((s) => s.prefs.notify);
  const hosting = useStore((s) => s.daemonStatus?.mode === "hosting");
  const announced = useRef(new Map<SessionId, AttentionReason>());

  useEffect(() => {
    const { out, next } = diffQueue(announced.current, queue, (id) => {
      const s = sessions[id];
      return s ? sessionLabel(s) : id.slice(0, 8);
    });
    announced.current = next;

    if (!enabled || !isTauri()) return;
    // Only the hosting window speaks — see the note at the top of this file.
    if (!hosting) return;
    // Not while you are looking at it. The queue is chrome: it is already on
    // screen, already re-ranked, and a banner for what you can see is the
    // noise that teaches people to dismiss banners.
    if (typeof document !== "undefined" && document.hasFocus()) return;

    for (const n of out) void deliver(n);
  }, [queue, sessions, enabled, hosting]);
}
